hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/impl/MutableStateBase.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (C) 2022 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hedera.node.app.state.impl;
+
+import com.hedera.node.app.spi.state.WritableState;
+import com.swirlds.common.merkle.MerkleNode;
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+import java.util.*;
+
+/**
+ * A base class for implementations of {@link WritableState}.
+ *
+ * @param <K> The key type
+ * @param <V> The value type
+ */
+abstract class MutableStateBase<K, V> extends StateBase<K, V> implements WritableState<K, V> {
```

Comment:
Just wondering, if modified and removed keys should also be listed when `readKeys()` is called.

Reply:
I've rewritten these classes and I think I cover all these cases now. Thanks for these insights!